syntax = "proto3";

// import "nanopb.proto";

// package rem;

option go_package = "rem" ;

enum ForwardingType {
    TO_ROOT = 0;
    RTN_HEADER_QUEUE = 1;

    TO_PARENT_TO_ROOT = 3;
    TO_PARENT = 4;
    TO_CHILD = 5;
    TO_NEIGHBORS = 6;
    TO_NODE=7;

    NEIGHBOR_STOP = 10;
    NEIGHBOR_TO_ROOT = 11;
    NEIGHBOR_AND_BACK = 12;
    PARENT_AND_BACK = 22;
}

// https://jpa.kapsi.fi/nanopb/docs/reference.html#proto-file-options
// int_size	Override the integer type of a field. (To use e.g. uint8_t to save RAM.)

enum HeaderType {
    BASIC = 0;
    DATA_BYTE = 1;
    LITTLE_BIG_ENDIAN_SYNC=15;
}

enum ProtobufType {
    NO_PROTOBUF = 0;

    PING = 1;
    TIMESYNC = 2;
    PONG = 3;
    MESH_TOPOLOGY = 4;
    INFO = 5;
    LOG = 6;

    SYNC_LITTLE_BIG_ENDIAN=15;
}

enum LoggingLevel {
    trace = 0 ;
    debug = 1 ;
    info = 2 ;
    warn = 3 ;
    error = 4 ;
    fatal = 5 ;
}

message RespirMeshInfo {
    uint32 target_id = 1;
    uint32 source_id = 2;
    ProtobufType type = 3;
}

message RespirMeshLog {
    uint32 source_time = 1;
    uint32 source_id = 2;
    // ProtobufType type = 3;
    LoggingLevel log_level = 4;
    string tags = 5;
    string message = 6;
}

message RespirMeshTimeSync {
    RespirMeshInfo info = 1 ;
    uint32 request_sent_time = 2;
    uint32 response_arrive_time = 3;
    uint32 response_sent_time = 4;
    uint32 request_arrive_time = 5;
    uint32 server_timestamp_unix = 6;
  }
